src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/FeatureSwitchManager.cs

```diff
@@ -472,7 +502,47 @@ public MethodIL GetMethodILWithInlinedSubstitutions(MethodIL method)
                 debugInfo = new SubstitutedDebugInformation(debugInfo, sequencePoints.ToArray());
             }
 
-            return new SubstitutedMethodIL(method, newBody, newEHRegions.ToArray(), debugInfo);
+            // We only optimize EcmaMethods because there we can find out the highest string token RID
+            // in use.
+            ArrayBuilder<string> newStrings = default;
+            if (hasGetResourceStringCall && method.GetMethodILDefinition() is EcmaMethodIL ecmaMethodIL)
+            {
+                // We're going to inject new string tokens. Start where the last token of the module left off.
+                // We don't need this token to be globally unique because all token resolution happens in the context
+                // of a MethodIL and we're making a new one here. It just has to be unique to the MethodIL.
+                int tokenRid = ecmaMethodIL.Module.MetadataReader.GetHeapSize(HeapIndex.UserString);
```

Comment:
Consider:
The current design will make it relatively confusing debugging the IL - as it will look like normal RID, but it won't be anywhere to find and different methods will have the same RIDs to make it even more confusing.
Another approach would be to generate the RIDs backwards - start with 0xFFFFFE and decrement (overflow if we hit the last valid RID in the module). This would not solve the duplicate RIDs across methods, but it would make the RIDs clearly special in some way. There's no need for the RIDs to be in one continuous block as far as I can tell.

Reply:
I had the 0xFFFFFE approach in the previous commit. I reverted that part because the code was harder to follow.

The rules in the compiler are that if you have an `EcmaMethodIL` class instance, the tokens in the IL match the tokens in the associated assembly because you're dealing with a real IL. If it's any other kind of `MethodIL`, all bets are off. This is constructing a special SubstitutedMethodIL. We don't count from top to bottom for the other synthetic method ILs. We could, it just didn't feel like a problem so far.